src/stash/src/postgres.rs

```diff
@@ -92,50 +92,48 @@ struct PreparedStatements {
-        let iter_key = client
-            .prepare(
-                "SELECT value, time, diff FROM data
+        let fetch_epoch_fut = client.prepare(match mode {
+            TransactionMode::Readonly | TransactionMode::Savepoint => {
+                // For readonly and savepoint stashes, don't attempt to
+                // increment the version and instead hard code it to 0 which
+                // will always fail the version check, since the version
+                // starts at 1 and goes up. Savepoint however will never
+                // retry COMMITs (and otherwise they'd retry forever because
+                // 0 will never succeed). Readonly can safely retry the
+                // original transaction.
+                "SELECT epoch, nonce, 0 AS version FROM fence"
+            }
+            TransactionMode::Writeable => {
+                "UPDATE fence SET version=version+1 RETURNING epoch, nonce, version"
+            }
+        });
+        let iter_key_fut = client.prepare(
+            "SELECT value, time, diff FROM data
              WHERE collection_id = $1 AND key = $2",
-            )
-            .await?;
-        let since = client
-            .prepare("SELECT since FROM sinces WHERE collection_id = $1")
-            .await?;
-        let upper = client
-            .prepare("SELECT upper FROM uppers WHERE collection_id = $1")
-            .await?;
-        let collection = client
-            .prepare("SELECT collection_id FROM collections WHERE name = $1")
-            .await?;
-        let iter = client
-            .prepare(
-                "SELECT key, value, time, diff FROM data
+        );
+        let since_fut = client.prepare("SELECT since FROM sinces WHERE collection_id = $1");
+        let upper_fut = client.prepare("SELECT upper FROM uppers WHERE collection_id = $1");
+        let collection_fut =
+            client.prepare("SELECT collection_id FROM collections WHERE name = $1");
+        let iter_fut = client.prepare(
+            "SELECT key, value, time, diff FROM data
              WHERE collection_id = $1",
-            )
-            .await?;
-        let seal = client
-            .prepare("UPDATE uppers SET upper = $1 WHERE collection_id = $2")
-            .await?;
-        let compact = client
-            .prepare("UPDATE sinces SET since = $1 WHERE collection_id = $2")
-            .await?;
+        );
+        let seal_fut = client.prepare("UPDATE uppers SET upper = $1 WHERE collection_id = $2");
+        let compact_fut = client.prepare("UPDATE sinces SET since = $1 WHERE collection_id = $2");
+
+        // Run all of our prepare statements in parallel.
+        let (fetch_epoch, iter_key, since, upper, collection, iter, seal, compact) = futures::try_join!(
+            fetch_epoch_fut,
+            iter_key_fut,
+            since_fut,
+            upper_fut,
+            collection_fut,
+            iter_fut,
+            seal_fut,
+            compact_fut
+        )?;
```

Comment:
Do you have any numbers for the type of speed up we get from this? If these all use the same underlying PostgreSQL client, then will rust-postgres end up serializing these behind the scenes?

I tried briefly looking at the `prepare` implementation. It looks like it sends a message over some channel and then a central event loop reads the messages off and processes them one-by-one.

Reply:
Yes, I think that's a great idea!